### PR TITLE
Add ballerina transaction support in RabbitMQ connector

### PR DESCRIPTION
**Pull request by aashikam**
## Purpose
Allow ballerina transactions in RabbitMQ client and service

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [x] API documentation 
  - [ ] Module documentation in Module.md files
  - [x] Ballerina By Examples

### CODE REVIEW

File: examples/rabbitmq-consumer/rabbitmq_consumer.description [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// In this example, the messages are consumed from an
+// already existing queue using the Ballerina RabbitMQ message listener.
```
Comment (praneesha) [inline examples/rabbitmq-consumer/rabbitmq_consumer.description]:
```suggestion
// existing queue using the Ballerina RabbitMQ message listener.
```

File: examples/rabbitmq-consumer/rabbitmq_consumer.description [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+// Ballerina RabbitMQ connection used here can be re-used to create
```
Comment (praneesha) [inline examples/rabbitmq-consumer/rabbitmq_consumer.description]:
```suggestion
// The Ballerina RabbitMQ connection used here can be re-used to create
```

File: examples/rabbitmq-consumer/rabbitmq_consumer.description [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+// multiple channels.
+// Multiple services, consuming messages from the same queue or from
```
Comment (praneesha) [inline examples/rabbitmq-consumer/rabbitmq_consumer.description]:
```suggestion
// Multiple services consuming messages from the same queue or from
```

File: examples/rabbitmq-consumer/rabbitmq_consumer.description [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+// different queues; can be attached to the same Listener.
```
Comment (praneesha) [inline examples/rabbitmq-consumer/rabbitmq_consumer.description]:
```suggestion
// different queues can be attached to the same Listener.
```

File: examples/rabbitmq-consumer/rabbitmq_consumer.out [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# Make sure to have the RabbitMQ broker running.
+#
+# Navigate to the directory that contains the
+# 'rabbitmq_consumer.bal' file and issue the 'ballerina run' command as follows..
```
Comment (praneesha) [inline examples/rabbitmq-consumer/rabbitmq_consumer.out]:
```suggestion
# 'rabbitmq_consumer.bal' file, and issue the 'ballerina run' command as follows.
```

File: examples/rabbitmq-producer/rabbitmq_producer.description [hunk 6]
```diff
@@ -0,0 +1,2 @@
+// In this example, messages are sent to two different queues;
```
Comment (praneesha) [inline examples/rabbitmq-producer/rabbitmq_producer.description]:
```suggestion
// In this example, messages are sent to two different queues,
```

File: examples/rabbitmq-producer/rabbitmq_producer.description [hunk 7]
```diff
@@ -0,0 +1,2 @@
...
+// using the same channel and using two different channels.
```
Comment (praneesha) [inline examples/rabbitmq-producer/rabbitmq_producer.description]:
```suggestion
// to one queue using the same channel and to the other using two different channels.
```

File: examples/rabbitmq-producer/rabbitmq_producer.out [hunk 8]
```diff
@@ -0,0 +1,6 @@
+# Make sure to have the RabbitMQ broker running.
+#
+# Navigate to the directory that contains the
+# 'rabbitmq_producer.bal' file and issue the 'ballerina run' command as follows..
```
Comment (praneesha) [inline examples/rabbitmq-producer/rabbitmq_producer.out]:
```suggestion
# 'rabbitmq_producer.bal' file, and issue the 'ballerina run' command as follows.
```

File: examples/rabbitmq-transaction-consumer/rabbitmq_transaction_consumer.description [hunk 9]
```diff
@@ -0,0 +1,5 @@
+// In this example, the messages are consumed from an
+// already existing queue using the Ballerina RabbitMQ message listener
```
Comment (praneesha) [inline examples/rabbitmq-transaction-consumer/rabbitmq_transaction_consumer.description]:
```suggestion
// existing queue using the Ballerina RabbitMQ message listener
```

File: examples/rabbitmq-transaction-consumer/rabbitmq_transaction_consumer.description [hunk 10]
```diff
@@ -0,0 +1,5 @@
...
+// and ballerina transactions.
```
Comment (praneesha) [inline examples/rabbitmq-transaction-consumer/rabbitmq_transaction_consumer.description]:
```suggestion
// and Ballerina transactions.
```

File: examples/rabbitmq-transaction-consumer/rabbitmq_transaction_consumer.description [hunk 11]
```diff
@@ -0,0 +1,5 @@
...
+// Upon successful execution of the transaction block,
+// the channel will commit and rollback in-case of any error.
```
Comment (praneesha) [inline examples/rabbitmq-transaction-consumer/rabbitmq_transaction_consumer.description]:
```suggestion
// the channel will commit, or rollback in the case of any error.
```

File: examples/rabbitmq-transaction-consumer/rabbitmq_transaction_consumer.out [hunk 12]
```diff
@@ -0,0 +1,12 @@
+# Make sure to have the RabbitMQ broker running.
+#
+# Navigate to the directory that contains the
+# 'rabbitmq_transaction_consumer.bal' file and issue the 'ballerina run' command as follows..
```
Comment (praneesha) [inline examples/rabbitmq-transaction-consumer/rabbitmq_transaction_consumer.out]:
```suggestion
# 'rabbitmq_transaction_consumer.bal' file, and issue the 'ballerina run' command as follows.
```

File: examples/rabbitmq-transaction-producer/rabbitmq_transaction_producer.description [hunk 13]
```diff
@@ -0,0 +1,4 @@
+// In this example, a message is sent to an existing queue
+// using Ballerina RabbitMQ channel and ballerina transactions.
```
Comment (praneesha) [inline examples/rabbitmq-transaction-producer/rabbitmq_transaction_producer.description]:
```suggestion
// using the Ballerina RabbitMQ channel and Ballerina transactions.
```

File: examples/rabbitmq-transaction-producer/rabbitmq_transaction_producer.description [hunk 14]
```diff
@@ -0,0 +1,4 @@
...
+// Upon successful execution of the transaction block,
+// the channel will commit and rollback in-case of any error.
```
Comment (praneesha) [inline examples/rabbitmq-transaction-producer/rabbitmq_transaction_producer.description]:
```suggestion
// the channel will commit, or rollback in the case of any error.
```

File: examples/rabbitmq-transaction-producer/rabbitmq_transaction_producer.out [hunk 15]
```diff
@@ -0,0 +1,11 @@
+# Make sure to have the RabbitMQ broker running.
+#
+# Navigate to the directory that contains the
+# 'rabbitmq_transaction_producer.bal' file and issue the 'ballerina run' command as follows..
```
Comment (praneesha) [inline examples/rabbitmq-transaction-producer/rabbitmq_transaction_producer.out]:
```suggestion
# 'rabbitmq_transaction_producer.bal' file, and issue the 'ballerina run' command as follows.
```

File: examples/rabbitmq-transaction-producer/rabbitmq_transaction_producer.out [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+
+$ ballerina run rabbitmq_transaction_producer.bal
+
+2019-06-07 11:44:46,695 INFO  [ballerina/transactions] - Created transaction: ee1fd05e-0fb3-408f-b6fd-1fd740a2cf31
+The queue was created successfully
```
Comment (praneesha) [inline examples/rabbitmq-transaction-producer/rabbitmq_transaction_producer.out]:
```suggestion
The queue was created successfully.
```

File: examples/rabbitmq-transaction-producer/rabbitmq_transaction_producer.out [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+The message was sent successfully
```
Comment (praneesha) [inline examples/rabbitmq-transaction-producer/rabbitmq_transaction_producer.out]:
```suggestion
The message was sent successfully.
```

File: examples/rabbitmq-simple-producer/rabbitmq_simple_producer.bal [hunk 18]
```diff
@@ -1,26 +1,21 @@
...
+        } else {
+            io:println("The queue was created successfully");
+        }
+        // Publishes the message using the routing key named "MyQueue".
+        var sendResult = newChannel->basicPublish("Hello from Ballerina", "MyQueue");
+        if (sendResult is error) {
+            io:println("An error occurred while sending the message");
+        } else {
+            io:println("The message was sent successfully");
+        }
```
Comment (daneshk) [inline examples/rabbitmq-simple-producer/rabbitmq_simple_producer.bal]: Formatting issue. Please check

File: examples/rabbitmq-transaction-consumer/rabbitmq_transaction_consumer.bal [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+    queueConfig: {
+        queueName: "MyQueue"
+    }
+}
+// Attaches the service to the listener.
+service testSimpleConsumer on new rabbitmq:ChannelListener({ host: "localhost", port: 5672 }) {
+
+    // Gets triggered when a message is received by the queue.
+    resource function onMessage(rabbitmq:Message message) {
+        transaction{
```
Comment (daneshk) [inline examples/rabbitmq-transaction-consumer/rabbitmq_transaction_consumer.bal]:
```suggestion
        transaction {
```

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/RabbitMQTransactionContext.java [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    @Override
+    public void close() {
+        // Do nothing
+    }
+
+    @Override
+    public XAResource getXAResource() {
+        // Not supported
```
Comment (wggihan) [inline stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/RabbitMQTransactionContext.java]: We should return a runtime exception here saying not supported.

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/RabbitMQTransactionContext.java [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+    public void handleTransactionBlock(Context context) {
+        boolean isInTransaction = context.isInTransaction();
+        if (isInTransaction) {
+            TransactionLocalContext transactionLocalContext = context.getLocalTransactionInfo();
+            BallerinaTransactionContext txContext = transactionLocalContext.getTransactionContext(connectorId);
+            if (Objects.isNull(txContext)) {
+                try {
+                    channel.txSelect();
+                } catch (IOException exception) {
+                    throw new RabbitMQConnectorException("I/O Error occurred while initiating the transaction"
```
Comment (wggihan) [inline stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/RabbitMQTransactionContext.java]:
```suggestion
                    throw new RabbitMQConnectorException("I/O Error occurred while initiating the transaction "
```

File: examples/rabbitmq-producer/rabbitmq_producer.bal [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+    // Publishing messages to different routing keys using the same channel.
+    // Publishes the message using newChannel1 to a different routing key named MyQueue2.
+    worker w3 {
+        var sendResult = newChannel1->basicPublish("Hello from Ballerina", "MyQueue2");
+        if (sendResult is error) {
+            io:println("An error occurred while sending the message to
+                    MyQueue2 using newChannel1");
+        }
+    }
+    _ = wait {w1, w2, w3};
```
Comment (riyafa) [inline examples/rabbitmq-producer/rabbitmq_producer.bal]: Is it an error that we are ignoring? If it is an error shall we handle/log it?

File: examples/rabbitmq-producer/rabbitmq_producer.bal [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+
+    // Declares the queue, MyQueue2.
+    var queueResult2 = newChannel2->queueDeclare(queueConfig = { queueName: "MyQueue2" });
+    if (queueResult2 is error) {
+        io:println("An error occurred while creating the queue, MyQueue2");
+    }
+
+    // Publishing messages to an exchange using a routing key.
+    // Publishes the message using newChannel1 and the routing key named MyQueue1.
+    worker w1 {
```
Comment (riyafa) [inline examples/rabbitmq-producer/rabbitmq_producer.bal]: Shall we also mention in the example why we need to publish using different workers?

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/RabbitMQTransactionContext.java [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+     */
+    public RabbitMQTransactionContext(BMap<String, BValue> channelObject, String connectorId) {
+        channel = (Channel) channelObject.getNativeData(RabbitMQConstants.CHANNEL_NATIVE_OBJECT);
+        this.connectorId = connectorId;
+    }
+
+    @Override
+    public void commit() {
+        try {
+            if (deliveryTag != -1 && !autoAck) {
```
Comment (riyafa) [inline stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/RabbitMQTransactionContext.java]: Shall we use a constant to represent the meaning of this numerical value? Shall we check in all places?

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/RabbitMQConstants.java [hunk 25]
```diff
@@ -104,6 +106,11 @@
     // Warning suppression
     public static final String UNCHECKED = "unchecked";
 
+    // Transaction constant fields
+    public static final String RABBITMQ_TRANSACTION_CONTEXT = "rabbitmq_transactional_context";
+    static final String COMMIT_FAILED = "Transaction commit failed ";
```
Comment (daneshk) [inline stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/RabbitMQConstants.java]:
```suggestion
    static final String COMMIT_FAILED = "Transaction commit failed. ";
```

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/RabbitMQConstants.java [hunk 26]
```diff
@@ -104,6 +106,11 @@
...
+    static final String ROLLBACK_FAILED = "Transaction rollback failed";
```
Comment (daneshk) [inline stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/RabbitMQConstants.java]:
```suggestion
    static final String ROLLBACK_FAILED = "Transaction rollback failed. ";
```

File: examples/rabbitmq-producer/rabbitmq_producer.bal (hunk 22 above)
Comment (aashikam) [reply examples/rabbitmq-producer/rabbitmq_producer.bal]: Resolved

File: examples/rabbitmq-producer/rabbitmq_producer.bal (hunk 23 above)
Comment (aashikam) [reply examples/rabbitmq-producer/rabbitmq_producer.bal]: Resolved

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/RabbitMQTransactionContext.java (hunk 20 above)
Comment (aashikam) [reply stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/RabbitMQTransactionContext.java]: Resolved

**Pull request merged**